pkg/mimir/modules.go

```diff
@@ -542,23 +545,65 @@ func (t *Mimir) initRuler() (serv services.Service, err error) {
 
 	t.Cfg.Ruler.Ring.KVStore.Multi.ConfigProvider = multiClientRuntimeConfigChannel(t.RuntimeConfig)
 	t.Cfg.Ruler.Ring.ListenPort = t.Cfg.Server.GRPCListenPort
-	rulerRegisterer := prometheus.WrapRegistererWith(prometheus.Labels{"engine": "ruler"}, prometheus.DefaultRegisterer)
-	var queryable, federatedQueryable prom_storage.Queryable
-	// TODO: Consider wrapping logger to differentiate from querier module logger
-	queryable, _, eng := querier.New(t.Cfg.Querier, t.Overrides, t.Distributor, t.StoreQueryables, rulerRegisterer, util_log.Logger, t.ActivityTracker)
-
-	if t.Cfg.Ruler.TenantFederation.Enabled {
-		if !t.Cfg.TenantFederation.Enabled {
-			return nil, errors.New("-ruler.tenant-federation.enabled=true requires -tenant-federation.enabled=true")
+
+	var embeddedQueryable prom_storage.Queryable
+	var queryFunc rules.QueryFunc
+
+	if len(t.Cfg.Ruler.Querier.Address) > 0 {
+		querierTransport, err := remote.NewTransport(t.Cfg.Ruler.Querier)
+		if err != nil {
+			return nil, err
 		}
-		// Setting bypassForSingleQuerier=false forces `tenantfederation.NewQueryable` to add
-		// the `__tenant_id__` label on all metrics regardless if they're for a single tenant or multiple tenants.
-		// This makes this label more consistent and hopefully less confusing to users.
-		const bypassForSingleQuerier = false
+		remoteQuerier := remote.New(querierTransport, t.Cfg.API.PrometheusHTTPPrefix, util_log.Logger, ruler.WithOrgIDHeader)
```

Comment:
I think we're missing case when tenant-federation and remote querying. Query-frontend can actually run queries for multiple tenants just fine, if they are concatenated with `|` (and query-frontend has correct flag enabled), so we could reuse `ruler.TenantFederationQueryFunc` and just pass same `queryFunc` in both parameters. WDYT?

Reply:
> so we could reuse `ruler.TenantFederationQueryFunc` and just pass same queryFunc in both parameters.

Not sure if I fully understood here, but having a look at `ruler.TenantFederationQueryFunc` implementation, that would be the same as simply passing queryFunc, right? 

Also note that remote evaluation for federated rules is already supported with the default remote querier, as it will strictly depend on `X-Scope-OrgID` [set](https://github.com/grafana/mimir/blob/fc5611424238fc70881922e0dd849aa94d62135f/pkg/ruler/querier.go#L334) for the outgoing request. In fact `TestFederatedRules` e2e test [has been adapted](https://github.com/grafana/mimir/blob/fc5611424238fc70881922e0dd849aa94d62135f/integration/ruler_test.go#L750-L753) to support remote evaluation and it works as expected. However, I agree that it is better to define a specific e2e test for this feature. So I'm going with it. 👍

Reply:
> > so we could reuse `ruler.TenantFederationQueryFunc` and just pass same queryFunc in both parameters.
> 
> Not sure if I fully understood here, but having a look at `ruler.TenantFederationQueryFunc` implementation, that would be the same as simply passing queryFunc, right?

I don't think it's exactly the same. Reason is that `ruler.TenantFederationQueryFunc` first sets tenant ID into the context (based on what `FederatedGroupContextFunc` sets), and only then calls `queryFunc`. Without `ruler.TenantFederationQueryFunc` there will not be correct tenant ID in the context set.

Reply:
Ah, I see that `WithOrgIDHeader` uses same `ruler.ExtractTenantIDs` to get the right tenant.

Looking at the implementation, I wonder if all that works correctly if group had no source tenants.

Reply:
Looking at it again, you're right that all the pieces are there for tenant federation and remote querier to work together correctly: `WithOrgIDHeader` will get the right tenant and set it into X-Scope-OrgID for query-frontend to use.

Reply:
new integration test available @ 49561c3e680a3e8cae55d3af6173a39ea921d5d0